l20n/migrate/context.py

```diff
@@ -0,0 +1,210 @@
+          message order, comments and sections.
+
+        - The current FTL files for the given language.
+
+        - The legacy (DTD, properties) translation files for the given
+          language.  The translations from these files will be transformed
+          into FTL and merged into the existing FTL files for this language.
+
+        - A list of `FTL.Entity` objects whose some nodes are special operation
+          nodes: CONCAT, COPY, EXTERNAL, PLURALS, REPLACE, SOURCE.
+    """
+
+    def __init__(self, lang, reference_dir, localization_dir):
+        self.ftl_parser = FTLParser()
+        self.ftl_serializer = FTLSerializer()
+
+        # An iterable of plural category names relevant to the context's
+        # language.  E.g. ('one', 'other') for English.
+        self.plural_categories = get_plural_categories(lang)
+
+        # Paths to directories with input data, relative to CWD.
+        self.reference_dir = reference_dir
+        self.localization_dir = localization_dir
+
+        # Parsed input resources stored by resource path.
+        self.reference_resources = {}
+        self.localization_resources = {}
+
+        # An iterable of `FTL.Entity` objects whose some nodes can be the
+        # transform operations.
+        self.transforms = {}
+
+        # A dict whose keys are `(path, key)` tuples corresponding to target
+        # FTL translations, and values are sets of `(path, key)` tuples
+        # corresponding to localized entities which will be migrated.
+        self.dependencies = {}
+
+    def read_ftl_resource(self, path):
+        f = codecs.open(path, 'r', 'utf8')
+        try:
+            contents = f.read()
+        except UnicodeDecodeError, err:
+            logger = logging.getLogger('migrate')
+            logger.error(u'Error reading file {}: {}'.format(path, str(err)))
+        f.close()
+
+        ast, errors = self.ftl_parser.parse(contents)
+
+        if len(errors):
+            logger = logging.getLogger('migrate')
+            for err in errors:
+                logger.error(u'Syntax error in {}: {}'.format(path, str(err)))
+
+        return ast
+
+    def add_reference(self, path, realpath=None):
+        """Add an FTL AST to this context's reference resources."""
+        fullpath = os.path.join(self.reference_dir, realpath or path)
+        self.reference_resources[path] = self.read_ftl_resource(fullpath)
+
+    def add_localization(self, path):
+        """Add an existing localization resource.
+
+        If it's an FTL resource, add an FTL AST.  Otherwise, it's a legacy
+        resource.  Use a compare-locales parser to create a dict of (key,
+        string value) tuples.
+        """
+        fullpath = os.path.join(self.localization_dir, path)
+        if fullpath.endswith('.ftl'):
+            ast = self.read_ftl_resource(fullpath)
+            self.localization_resources[path] = ast
+        else:
+            parser = getParser(fullpath)
+            parser.readFile(fullpath)
```

Comment:
Nope, we'll need to convert partial localizations. Think of a feature with some translations in the DTD but the properties file is completely missing.